src/db/spanner/models.rs

```diff
@@ -0,0 +1,1292 @@
+    /// The "current time" on the server used for this session's operations
+    timestamp: SyncTimestamp,
+    /// Cache of collection modified timestamps per (user_id, collection_id)
+    coll_modified_cache: HashMap<(u32, i32), SyncTimestamp>,
+    /// Currently locked collections
+    coll_locks: HashMap<(u32, i32), CollectionLock>,
+}
+
+#[derive(Clone, Debug)]
+pub struct SpannerDb {
+    pub(super) inner: Arc<SpannerDbInner>,
+
+    /// Pool level cache of collection_ids and their names
+    coll_cache: Arc<CollectionCache>,
+}
+
+pub struct SpannerDbInner {
+    pub(super) conn: Conn,
+
+    thread_pool: Arc<::tokio_threadpool::ThreadPool>,
+    session: RefCell<SpannerDbSession>,
+}
+
+impl fmt::Debug for SpannerDbInner {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "SpannerDbInner")
+    }
+}
+
+impl Deref for SpannerDb {
+    type Target = SpannerDbInner;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+macro_rules! batch_db_method {
+    ($name:ident, $batch_name:ident, $type:ident) => {
+        pub fn $name(&self, params: params::$type) -> Result<results::$type> {
+            batch::$batch_name(self, params)
+        }
+    }
+}
+
+impl SpannerDb {
+    pub fn new(
+        conn: Conn,
+        thread_pool: Arc<::tokio_threadpool::ThreadPool>,
+        coll_cache: Arc<CollectionCache>,
+    ) -> Self {
+        let inner = SpannerDbInner {
+            conn,
+            thread_pool,
+            session: RefCell::new(Default::default()),
+        };
+        SpannerDb {
+            inner: Arc::new(inner),
+            coll_cache,
+        }
+    }
+
+    pub(super) fn get_collection_id(&self, name: &str) -> Result<i32> {
+        if let Some(id) = self.coll_cache.get_id(name)? {
+            return Ok(id);
+        }
+        let spanner = &self.conn;
+
+        let mut sql = ExecuteSqlRequest::default();
+        sql.sql = Some("SELECT id FROM collections WHERE name = @name;".to_string());
+        let mut params = HashMap::new();
+        params.insert("name".to_string(), name.to_string());
+        sql.params = Some(params);
+
+        let session = spanner.session.name.as_ref().unwrap();
+        let result = spanner
+            .hub
+            .projects()
+            .instances_databases_sessions_execute_sql(sql, session)
+            .doit();
+        let rv = match result {
+            Err(e) => {
+                println!("err {}", e);
+                // TODO Return error
+                Ok(0)
+            }
+            Ok(res) => {
+                println!("ok {:?}", res.1);
+                match res.1.rows {
+                    Some(row) => Ok(row[0][0].parse().unwrap()),
+                    None => Ok(0),
+                }
+            }
+        };
+        if let Ok(val) = rv {
+            self.coll_cache.put(val, name.to_owned())?;
+        };
+        rv
+        //let id = sql_query("SELECT id FROM collections WHERE name = ?")
```

Comment:
Is this needed? Could we move it to a declared comment and add a bit of context to prevent it from being dropped as "dead code" later?

Reply:
I had intended to delete these blocks actually, thanks for the reminder. I'll open a followup bug for it.